### ISSUE TITLE
Gradient shape error.

### ISSUE THREAD

**Issue by sschoenholz** [google/jax#1350]
Consider the following code,

```python
def pairwise_displacement(Ra, Rb):
  return Ra - Rb

def periodic_displacement(side, dR):
  return np.mod(dR + side * f32(0.5), side) - f32(0.5) * side

def square_distance(dR):
  return np.sum(dR ** 2, axis=-1)


def distance(dR):
  return np.sqrt(square_distance(dR))
  
def periodic(side):
  def displacement_fn(Ra, Rb):
    return periodic_displacement(side, pairwise_displacement(Ra, Rb))
  return displacement_fn


def get_samples(sigma):
  particle_count = 10
  key = random.PRNGKey(0)

  N_2 = particle_count // 2
  particle_volume = N_2 * np.pi * (sigma ** 2 + 1.) / 4.
  box_size = np.sqrt(particle_volume / 0.96)
  
  displacement = periodic(box_size)

  R = box_size * random.uniform(key, (particle_count, 2), dtype=f64)
  displacement = vmap(vmap(displacement, (None, 0)), (0, None))
  return np.sum(displacement(R, R))


print('Got shape {} but expected a scalar!!'.format(
    grad(get_samples)(1.0).shape))
```
The function `grad(get_samples)(s)` should return a scalar. However, we get something of shape `[particle_count, particle_count, 2]`. It seems like the issue is most likely coming from tracing through `periodic(box_size)` since without that line the shapes work out as expected.

**Comment by mattjj** [comment]
Some more minimization:

```python
import jax.numpy as np
from jax import vmap, grad
from jax import random

f64 = np.float64
f32 = np.float32

def pairwise_displacement(Ra, Rb):
  return Ra - Rb

def periodic_displacement(side, dR):
  # return dR + side  # NOTE this works
  return np.mod(dR, side)


def periodic(side):
  def displacement_fn(Ra, Rb):
    return periodic_displacement(side, pairwise_displacement(Ra, Rb))
  return displacement_fn


def get_samples(sigma):
  particle_count = 10
  R = np.zeros((particle_count, 2), dtype=f64)
  displacement = vmap(vmap(periodic(sigma), (None, 0)), (0, None))
  return np.sum(displacement(R, R))

print('Got shape {} but expected a scalar!!'.format(
    grad(get_samples)(1.0).shape))
```

Notice the `# NOTE this works` comment. That makes me think there's a broadcasting issue with `np.mod`...
